### ISSUE TITLE
Usage with JSON Patching?

### ISSUE THREAD

**Issue by skokenes** [calmm-js/partial.lenses#211]
First off, this is a really great project! I have a question on how it might be leveraged alongside JSON patching. Here is my scenario:

I have a client-side apps that all share a single JSON document structure. This JSON document is managed on a server. Any client can send patches for the JSON document to the server via WebSockets. Those patches get applied and the JSON document is updated for all connected clients via the WebSockets.

In a single client app, I want to do local manipulations of the data in this JSON document. Very similar to the example in the tutorial. The library works great for that! Using `get`s, `set`s, and composition, I can locally modify parts of my JSON.

Once the local manipulation is done, I need to inform the server about these changes with patches so that they can be replayed for other connected clients. I am not sure how to translate the operations that my lenses are performing locally into an equivalent JSON patch. For example, given the following lens:
```javascript
L.set(
  L.compose(L.prop('titles'), L.index(0), L.prop('text')),
  'New title',
  sampleTitles
)
```

I want to generate an equivalent patch like: 
```json
{
  "op": "replace",
  "path": "/titles/0/text",
  "value": "New Title"
}
```

Any thoughts or suggestions on strategies for doing this? I appreciate any feedback that you can provide.

**Comment by polytypic** [comment]
How about using a library like [jiff](https://github.com/cujojs/jiff)  to generate patches after arbitrary changes?  I've never used jiff myself, but it would seem to fit the job.  The library was mentioned on the [JSON Patch](http://jsonpatch.com/) site.

I assume you want to be able to perform local client side manipulations without limiting yourself to the operations specified by JSON patches.  However, if you are primarily dealing with just JSON patches, then it might be simpler to just use a JSON patch library to perform patches.

BTW, note that Partial Lenses also provides the [`L.pointer`](https://github.com/calmm-js/partial.lenses/#L-pointer) lens constructor.  It should be easy to provide some support for applying JSON patches.  Here is [a sketch of that](https://calmm-js.github.io/partial.lenses/playground.html#MYewdgzgLgBADgQysAFjAvDA3gKAEQhx4BcMeATgKZwA2CwleANPolCiWQPRQCWUNShC4AGHpQAeUZvgBuCGgFdGpPADlKAdxgAVfoLw4AvjhyhIsNqh0gAYorDA+4DPCSoMAPmw4Im-h4AFFYoAHSEAJQ+wAgQlGRUtPQqOFRQiuRgMAAyoXFQgblwILxgUJTkwe5hbCgRTG7IYfJKlBE4XFwwoT04ACaUAGYIijRQxDjs5CDaAKLk05UABlgh4XBGMGAgsACelLAQinDF5OV9S+0mJmbg0DAQCAC2tJQ6klAQrrh8AkKkAG0sHQwABzRQIUGUVSUMDMGDlKSqPR-PBGAC6xlMIRs9kczjAVSaEUCjxegneUggESAA).  One problem might be that JSON patch seems to have somewhat different semantics concerning failure, but I haven't thought about this in detail.

**Comment by skokenes** [comment]
Yea, an ideal situation is that I could do cool lens magic client side to manipulate my data and then just autogenerate the patches that need to be sent to the server based on that. My server already implements the JSON patch application piece so I don't have worry about that. There have been a few issues though. Here are the approaches that I've tried:

1. Using a JSON diff'ing library to generate the patches. I haven't tried jiff yet. With the libraries that I did try, they worked great for "replace" type operations but did not handle add/removes from arrays well, instead generating convoluted sequences of "replace" operations. This is a problem for my architecture because my server handles multi-user patching conflicts with operational transforms, and the sequence of "replace" patches is not the equivalent of an "add"/"remove" when looking at it from an OT perspective. At a glance, jiff looks like it does a better job of turning Array manipulations into "add"/"remove" patches so I will test that out!

2. Generate the patch by creating a JSON pointer  from the lens being used (via `tieIx` calls) and the value to be set. This also works really well for "replace" type operations (like a basic "set" to update a value), but there are edge cases where it can break: for `modify`, you'd have to get the lens at the resulting structure to get the final value. If the lens does any transforms on the data like sorting (like in the tutorial), then the generated JSON pointer won't actually match the source structure. Appends and removals also might be tricky to capture properly.

3. Just don't use lens setters and do all inserts/removes/updates via patching. This removes the ability to do modifications with lenses (reserving them for gets instead) but does give me finer control over the patches that I send to the server.

I was leaning towards option 3 until you suggested `jiff`, which may handle the array problems that I was having with other diff'ing libraries. So I'll take another look at 1.

Thanks for the feedback!
